pkg/cloudprovider/types.go

```diff
@@ -239,27 +262,74 @@ func (i InstanceTypeOverhead) Total() corev1.ResourceList {
 }
 
+// ReservationManager is used to track the availability of a reserved offering over the course of a scheduling
+// simulation. Reserved offerings may have a limited number of available instances associated with them,
+// This is exposed as an interface for cloudprovider's to implement to give flexibility when dealing with separate
+// offerings with associated availablility.
+type ReservationManager interface {
+	// Reserve takes a unique identifier for a reservation, and returns a boolean indicating if the reservation was
+	// successful. Reserve should be idempotent, i.e. multiple calls with the same reservation ID should only count for a
+	// single reservation.
+	Reserve(string) bool
+	// Release takes a unique identifier for a reservation, and should discard any matching reservations. If no
+	// reservations exist for the given id, release should be a no-op.
+	Release(string)
+}
+
 // An Offering describes where an InstanceType is available to be used, with the expectation that its properties
 // may be tightly coupled (e.g. the availability of an instance type in some zone is scoped to a capacity type) and
 // these properties are captured with labels in Requirements.
-// Requirements are required to contain the keys v1.CapacityTypeLabelKey and corev1.LabelTopologyZone
+// Requirements are required to contain the keys v1.CapacityTypeLabelKey and corev1.LabelTopologyZone.
 type Offering struct {
+	// ReservationManager is used for tracking availabity of reserved offerings over the course of a scheduling loop. It
+	// must be non-nil for offerings with capacity type "reserved", but may be nil otherwise.
+	ReservationManager
+
 	Requirements scheduling.Requirements
 	Price        float64
-	// Available is added so that Offerings can return all offerings that have ever existed for an instance type,
-	// so we can get historical pricing data for calculating savings in consolidation
-	Available bool
+	Available    bool
 }
 
 type Offerings []Offering
 
+// WithCapacityType filters the offerings by the provided capacity type.
+func (ofs Offerings) WithCapacityType(capacityType string) Offerings {
+	return lo.Filter(ofs, func(o Offering, _ int) bool {
+		return o.Requirements.Get(v1.CapacityTypeLabelKey).Any() == capacityType
+	})
+}
+
+// Reserve attempts to make a reservation for each offering, returning true if it was successful for any.
+func (ofs Offerings) Reserve(id string) Offerings {
+	return lo.Filter(ofs, func(o Offering, _ int) bool {
+		return o.Reserve(id)
+	})
+}
+
+func (ofs Offerings) Release(id string) {
+	for i := range ofs {
+		ofs[i].Release(id)
+	}
+}
+
 // Available filters the available offerings from the returned offerings
 func (ofs Offerings) Available() Offerings {
 	return lo.Filter(ofs, func(o Offering, _ int) bool {
 		return o.Available
 	})
 }
 
+func (ofs Offerings) PartitionCompatible(reqs scheduling.Requirements) (compatible Offerings, incompatible Offerings) {
```

Comment:
Same kind of question here about creating two new offerings arrays, I think this is going to end up being pretty inefficient in the scheduler at the end of the day